### ISSUE TITLE
[J11] Prepare Java 11 Support

### ISSUE THREAD

**Issue by cdietrich** [eclipse/xtext#1182]
Xtext should not "explode" with Java 11

- JavaVersion constant and its from and to conversions
    Wizard
- new Maven/Gradle Dependencies to JDT we consume needed
- new ASM needed? (https://dev.eclipse.org/ipzilla/show_bug.cgi?id=17016) https://github.com/eclipse/xtext/issues/1325 :ballot_box_with_check: 
- new Tycho Version needed? yes we need tycho 1.3 https://github.com/eclipse/xtext/issues/1322
- Gradle? gradle 4.10.2 should be fine. :ballot_box_with_check: 
- Maven? 
- new Guava?
- new Guice?
- Function wise (do not explode)

**Comment by cdietrich** [comment]
see https://github.com/eclipse/xtext-core/issues/517

**Comment by cdietrich** [comment]
see https://groups.google.com/forum/#!topic/xtend-lang/T2TLFWVl_C4

**Comment by cdietrich** [comment]
wont ship this with 2.15 so moved to 2.16

**Comment by jpaw** [comment]
An upgrade of guice will be required. As a quick test, I changed the guice release from 3.0 to 4.2.0. In xtext-core, then 12 tests fail (but that's less than 1% of all tests :-)).
In xtext-xtend, I also had to upgrade org.eclipse.core.runtime in order to avoid some signature mismatch exception, but afterwards, the resulting xtend-maven-plugin does not show the warnings about illegal reflective access any more.
(Btw, I used the "no_aop" classifier for guice, this did not make any difference related to the number of failing tests. guice still bundles / repackages cglib and asm, and 4.2.0 still does not use the most recent ones, but with the classifiers "no_aop" and "classes" versions without that are available as well.)

**Comment by cdietrich** [comment]
how did you deal with guice 4.x needing a newer guava? see https://github.com/eclipse/xtext-core/issues/393

**Comment by cdietrich** [comment]
@szarnekow do you know of any ussages of the guice aop part? do you remember which version of guice we use?

**Comment by cdietrich** [comment]
@jpaw did you investigate if this can be done without a guice/guava bump? for java 9 and 10 it was possible. or did you do the bump to get rid of the reflection warnings?

**Comment by svenefftinge** [comment]
We should not bump the major version. Also see discussion here: https://dev.eclipse.org/mhonarc/lists/modeling-pmc/msg04475.html

**Comment by jpaw** [comment]
Related to the version bump of guava:
guice used 19.0 until very recently, in commits eface22... (just before the 4.2.0 tag) and in 289073...
apparently mainly to be able to use a newer version of the truth test library (that upgrade was done after the 4.2.0 tag). This would allow the version range [19.0, 20.0 and 22.0) which should work for xtext as well as guice. I rebuilt guice 4.2.0, but with a guava 20.0 dependency, and same results (20.0 is the last guava release which offers Java 6 compatibility).

Related to guice: xtext currently uses 3.0, which does not yet offer the no_aop version. guice 3.x is officially not even compatible with Java 8 (guice 4.0 ff are), so I doubt it will work with Java 11.

With my test builds of the xtext / xtend plugins, I'm currently stuck with "ResourceNotLoaded" or
`
org.eclipse.emf.ecore.resource.impl.ResourceSetImpl$1DiagnosticWrappedException: org.eclipse.xtext.common.types.access.TypeResource$CannotLoadTypeResourceException: java.lang.UnsupportedOperationException
    at org.eclipse.emf.ecore.resource.impl.ResourceSetImpl.handleDemandLoadException (ResourceSetImpl.java:319)
    at org.eclipse.emf.ecore.resource.impl.ResourceSetImpl.demandLoadHelper (ResourceSetImpl.java:278)
    at org.eclipse.xtext.resource.XtextResourceSet.getResource (XtextResourceSet.java:265)
`
I will retest this with emf 2.15.0 as soon as that has been released (current test used emf 2.12.0).

**Comment by cdietrich** [comment]
it was working with java 9 and 10. the question: are we talking about the xtext built itself or projects using xtext?

**Comment by jpaw** [comment]
I'm currently mainly interested in using xtext, that is, invoking xtext-maven-plugin and xtend-maven-plugin from a maven build job which is executed by JDK 11. 
Of course the fallback would be to run the build with Java 8 and to only use Java 11 at runtime, but that would restrict using Java 9/10/11 features to be used in pure Java projects only.

**Comment by cdietrich** [comment]
I would have expected that would have worked by simply bumping the jdt version (not out yet) and introducing a new entry in the javaversion enum

**Comment by cdietrich** [comment]
we require asm7 and the usage of OpCodes.ASM7 to get this running

**Comment by cdietrich** [comment]
Address `javax.annotations.Generated`

**Comment by chrswk** [comment]
So Java 11 support is planned for 2.16.0, correct? Do you have any idea about the schedule?

**Comment by cdietrich** [comment]
2.16 is planned for 2018-12. We are still lining up for
- ASM 7 final at maven and eclipse orbit
- Tycho 1.3
- Maven Consumable Artifacts for JDT

**Comment by chrswk** [comment]
Great, thanks for the information! 👍

**Comment by cdietrich** [comment]
emf.codegen.ecore needs `org.eclipse.jdt.core.formatter.CodeFormatter` which is not part of `ecj`

**Comment by cdietrich** [comment]
we need to adress the removal of javax.annotation in tycho.
adding the following to mydsl.tests

```
<plugin>
				<groupId>org.eclipse.tycho</groupId>
				<artifactId>target-platform-configuration</artifactId>
				<configuration>
					<dependency-resolution>
						<extraRequirements>
							<!-- to get the org.eclipse.osgi.compatibility.state plugin
							if the target platform is Luna or later.
							(backward compatible with kepler and previous versions)
							see https://bugs.eclipse.org/bugs/show_bug.cgi?id=492149 -->
							<requirement>
								<type>eclipse-feature</type>
								<id>org.eclipse.rcp</id>
								<versionRange>0.0.0</versionRange>
							</requirement>
						</extraRequirements>
					</dependency-resolution>
				</configuration>
			</plugin>
```
seems to help

**Comment by cdietrich** [comment]
plus we might need to change the wizard to activate the modules depending on execution, not the selected java version to be able to build with java 11 against java 8

```
	<properties>
		<tycho.testArgLine>--add-modules=ALL-SYSTEM</tycho.testArgLine>
	</properties>
```

**Comment by cdietrich** [comment]
Moved to 2.17 since there is no maven consumable jdt artifacts available

**Comment by cdietrich** [comment]
TO be discussed: do we want to bump platform versions too

**Comment by cdietrich** [comment]
Fixed in Xtext/Xtend 2.17
